roles/docker-rootless/tasks/main.yml

```diff
@@ -0,0 +1,72 @@
+---
+
+- name: Ensure directory structure exists
+  file:
+    path: '{{ rootlessdocker_install_dir }}/{{ item.path }}'
+    owner: 'root'
+    group: 'root'
+    state: directory
+    mode: '0755'
+  with_filetree: '{{ role_path }}/templates/rootless-docker'
+  when: item.state == 'directory'
+
+- name: Copy over templates.
+  template:
+    src: '{{ item.src }}'
+    dest: '{{ rootlessdocker_install_dir }}/{{ item.path }}'
+    owner: 'root'
+    group: 'root'
+    mode: '{{ item.mode }}'
+    force: no
+  with_filetree: '{{ role_path }}/templates/rootless-docker'
+  when: item.state == 'file'
+
+- name: add rootless docker profile script to source modules
+  template:
+    src: '{{ item }}'
+    dest: '/etc/profile.d/{{ item }}'
+    owner: 'root'
+    group: 'root'
+    mode: '0644'
+    force: no
+  with_items:
+  - 'z96_rootlessdocker_modules.sh'
+
+- name: install rootless docker
+  shell: |
```

Comment:
This all needs to be written in Ansible as much as possible

Reply:
I cannot completely get away from running some sort of command/shell to run the downloaded rootless docker installation script.

I can refactor the inlined shell script with a separate task to use `get_url` instead of curl within the script. But that's one line and with a separate task it becomes several lines. I'll do that if that's a better convention.

The other parts are not so easy to implement directly with Ansible. The subuid/subgid logic is temporary otherwise the rootless docker install script will fail. It's simple to do with shell script, a couple of lines. Otherwise it's somewhat complicated with Ansible. I found an example here:
https://github.com/debops/debops/blob/d1ed2f31d9816637e773033982c33c511c870edc/ansible/roles/root_account/templates/etc/ansible/facts.d/root_account.fact.j2#L45

There's no builtin Ansible feature for subuid/subgid that I found (I did find a feature request for this functionality).

Reply:
I don't think this is better because it is more verbose and I have to keep track of tasks inter-dependencies. I can refactor to something like this (I haven't tested this):
```
- name: check if rootless docker already installed successfully
  stat: 
    path: "{{ rootlessdocker_install_dir }}/config/install_success"
  register: rootless_docker_installed

- name: Download rootless docker install script
  get_url:
    url="https://get.docker.com/rootless"
    dest=/tmp/
    mode: '0755'
  register: get_rootless_docker
  when: not rootless_docker_installed.stat.exists

- name: install rootless docker
  shell: |
    userid=$(id -u)
    largestuid=$(cat /etc/subuid | awk '{split($0,a,":"); print a[2]}' | sort -n | tail -1)
    mysubuid=$(( largestuid + 65536 ))
    cleansubuid=false
    if ! grep "^$(id -un):\|^$(id -u):" /etc/subuid >/dev/null 2>&1; then
      echo ${userid}:${mysubuid}:65536 >> /etc/subuid
      cleansubuid=true
    fi

    largestgid=$(cat /etc/subgid | awk '{split($0,a,":"); print a[2]}' | sort -n | tail -1)
    mysubgid=$(( largestgid + 65536 ))
    cleansubgid=false
    if ! grep "^$(id -un):\|^$(id -u):" /etc/subgid >/dev/null 2>&1; then
      echo ${userid}:${mysubgid}:65536 >> /etc/subgid
      cleansubgid=true
    fi

    FORCE_ROOTLESS_INSTALL=1 DOCKER_BIN="{{ rootlessdocker_install_dir }}/bin" /tmp/rootless

    if [ $? -eq 0 ] ; then
      touch "{{ rootlessdocker_install_dir }}/config/install_success"
    fi

    chown root:root {{ rootlessdocker_install_dir }}/bin/*

    if [ "$cleansubuid" = true ] ; then
      sed -i "/^${userid}:/d" /etc/subuid
    fi

    if [ "$cleansubgid" = true ] ; then
      sed -i "/^${userid}:/d" /etc/subgid
    fi
  args:
    executable: /bin/bash
    creates: "{{ rootlessdocker_install_dir }}/config/install_success"

- name: Clean up rootless docker script
  file:
    path: /tmp/rootless
    state: absent

```

I end up having to add three additional tasks just to avoid curl command in the shell command i.e. part of a single line:
```
    curl -fsSL https://get.docker.com/rootless | \
      FORCE_ROOTLESS_INSTALL=1 DOCKER_BIN="{{ rootlessdocker_install_dir }}/bin" sh
```

I have not written ansible scripts before, so I'm not very experienced. Perhaps someone could suggest a better implementation, but it just does not seem to be good code to make something so verbose when it can be a single line. If there was an ansible module to manipulate subuid/subgid, then using just ansible modules would be simpler. But I did not find such a module.